api/controllers/tag.go

```diff
@@ -0,0 +1,201 @@
+		TagStore: t,
+	}
+}
+
+func (t *TagController) Routes() []*fireball.Route {
+	return []*fireball.Route{
+		{
+			Path: "/tag",
+			Handlers: fireball.Handlers{
+				"GET":    t.ListTags,
+				"POST":   t.CreateTag,
+				"DELETE": t.DeleteTag,
+			},
+		},
+	}
+}
+
+func (t *TagController) DeleteTag(c *fireball.Context) (fireball.Response, error) {
+	var tag models.Tag
+	if err := json.NewDecoder(c.Request.Body).Decode(&tag); err != nil {
+		return nil, errors.New(errors.InvalidRequest, err)
+	}
+
+	if err := t.TagStore.Delete(tag.EntityType, tag.EntityID, tag.Key); err != nil {
+		return nil, err
+	}
+
+	return fireball.NewJSONResponse(200, nil)
+}
+
+func (t *TagController) CreateTag(c *fireball.Context) (fireball.Response, error) {
+	var tag models.Tag
+	if err := json.NewDecoder(c.Request.Body).Decode(&tag); err != nil {
+		return nil, errors.New(errors.InvalidRequest, err)
+	}
+
+	if err := t.TagStore.Insert(tag); err != nil {
+		return nil, err
+	}
+
+	return fireball.NewJSONResponse(200, tag)
+}
+
+func (t *TagController) ListTags(c *fireball.Context) (fireball.Response, error) {
+	query := c.Request.URL.Query()
+
+	tags, err := t.selectTagsFromQuery(query)
+	if err != nil {
+		return nil, err
+	}
+
+	entityTags := tags.GroupByID()
+	if environmentID := query.Get(client.TagQueryParamEnvironmentID); environmentID != "" {
+		filterTagsByEnvironmentID(entityTags, environmentID)
+	}
+
+	if name := query.Get(client.TagQueryParamName); name != "" {
+		filterTagsByName(entityTags, name)
+	}
+
+	if fuzz := query.Get(client.TagQueryParamFuzz); fuzz != "" {
+		filterTagsByIDOrNameGlob(entityTags, fuzz)
+	}
+
+	if version := query.Get(client.TagQueryParamVersion); version != "" {
+		if version == "latest" {
+			filterTagsByLatestVersion(entityTags)
+		} else {
+			filterTagsByVersion(entityTags, version)
+		}
+	}
+
+	filteredTags := models.Tags{}
+	for _, tags := range entityTags {
+		filteredTags = append(filteredTags, tags...)
+	}
+
+	return fireball.NewJSONResponse(200, filteredTags)
+}
+
+func (t *TagController) selectTagsFromQuery(query url.Values) (models.Tags, error) {
+	entityID := query.Get(client.TagQueryParamID)
+	entityType := query.Get(client.TagQueryParamType)
+
+	if entityType != "" && entityID != "" {
+		return t.TagStore.SelectByTypeAndID(entityType, entityID)
+	}
+
+	if entityType != "" {
+		return t.TagStore.SelectByType(entityType)
+	}
+
+	if entityID != "" {
+		tags, err := t.TagStore.SelectAll()
+		if err != nil {
+			return nil, err
+		}
+
+		return tags.WithID(entityID), nil
+	}
+
+	return t.TagStore.SelectAll()
+}
+
+func filterTagsByEnvironmentID(entityTags map[string]models.Tags, environmentID string) {
+	for entityID, tags := range entityTags {
+		hasMatch := tags.Any(func(t models.Tag) bool {
+			return t.Key == "environment_id" && t.Value == environmentID
+		})
+
+		if !hasMatch {
+			delete(entityTags, entityID)
+		}
+	}
+}
+
+func filterTagsByName(entityTags map[string]models.Tags, name string) {
```

Comment:
[nit] same as above - but I see there are a few methods which modify the argument passed in to filter. preferable in go as maps are actually pointers?